### ISSUE TITLE
Error generating mesh

### ISSUE THREAD

**Issue by PavelBlend** [doyubkim/fluid-engine-dev#289]
It seems to me that the polygon mesh generator does not work correctly at some points.
The mesh is not Manifold.
![01](https://user-images.githubusercontent.com/7983249/78382704-ee09d500-75df-11ea-8f57-4aae5a69ff05.jpg)
[test_apic_method.zip](https://github.com/doyubkim/fluid-engine-dev/files/4428565/test_apic_method.zip)

**Comment by doyubkim** [comment]
Looks like a marching cube problem for sure. Let me take a look.

**Comment by utilForever** [comment]
I'll check this issue. Thanks. :)

**Comment by PavelBlend** [comment]
I tested the simulator again and noticed one thing:
These invalid triangles are only created on the walls of the domain. I started to study the code in the file `src\jet\marching_cubes.cpp` to get a rough idea of the cause of the error. I saw that the mesh generator uses different algorithms for creating triangles for the walls of the domain and for the triangles within the domain. The error occurs in blocks of code that have comments:
`Construct boundaries parallel to x-y plane`
`Construct boundaries parallel to y-z plane`
`Construct boundaries parallel to x-z plane`
https://github.com/doyubkim/fluid-engine-dev/blob/45b4bdbdb4c6d8c0beebc682180469198203b0ef/src/jet/marching_cubes.cpp#L395
I hope this information will help you understand the cause of the error.

**Comment by doyubkim** [comment]
Thanks @PavelBlend ! So I guess what’s happening here is that boundary mesher is adding more triangles and the topology of it does not agree with the surface mesher. Maybe it’s direction-dependent; one side of the domain works fine but the opposite side might show this problem.  @utilForever: one way to fix this is to make a simple repro case with two volumes close together and see how the lookup table is being referred for the particular case.

**Comment by PavelBlend** [comment]
@doyubkim I created a small example:
```python
from pyjet import *
import numpy as np


ANIM_NUM_FRAMES = 360
ANIM_FPS = 60
Logging.mute()
# Create APIC solver
resX = 50
solver = ApicSolver3(resolution=(resX, resX, resX), domainSizeX=1.0)
solver.useCompressedLinearSystem = True
# Setup emitter
sphere = Sphere3(center=(0.5, 0.5, 0.5), radius=0.15)
emitter = VolumeParticleEmitter3(implicitSurface=sphere, spacing=1.0 / (2 * resX), isOneShot=True)
solver.particleEmitter = emitter
# Convert to surface
grid_size = 1.0 / resX
grid = VertexCenteredScalarGrid3((resX, resX, resX), (grid_size, grid_size, grid_size))
# Make first frame
frame = Frame(0, 1.0 / ANIM_FPS)
for i in range(ANIM_NUM_FRAMES):
    print('Frame {:d}'.format(i))
    solver.update(frame)
    pos = np.array(solver.particleSystemData.positions, copy=False)
    converter = SphPointsToImplicit3(1.5 * grid_size, 0.5)
    converter.convert(pos.tolist(), grid)
    surface_mesh = marchingCubes(
        grid,
        (grid_size, grid_size, grid_size),
        (0, 0, 0),
        0.0,
        DIRECTION_ALL,
        DIRECTION_NONE
    )
    surface_mesh.writeObj('frame_{:06d}.obj'.format(i))
    frame.advance()
```
at frame 48 I get this mesh:
![011111](https://user-images.githubusercontent.com/7983249/147454750-5565a58d-be21-4d40-9da0-635aea19d530.jpg)

Here is the obj file:
[frame_000048.zip](https://github.com/doyubkim/fluid-engine-dev/files/7779395/frame_000048.zip)

I think this is due to the fact that two drops are in the same voxel:
![01000](https://user-images.githubusercontent.com/7983249/147454918-814a0e95-2830-44bc-adce-62b2b384b38f.jpg)


The mesh generator builds a polygon between the drops.

**Comment by utilForever** [comment]
@PavelBlend Thanks!

**Comment by doyubkim** [comment]
Yup, I think this is definitely a marching square issue where its lookup table is not consistent with the marching cube.
